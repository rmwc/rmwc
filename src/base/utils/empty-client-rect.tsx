// as of mcw 14, mcw themselves has typed it as any
export const emptyClientRect: any = {
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0
};
