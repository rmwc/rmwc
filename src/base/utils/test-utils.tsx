import React from 'react';
import { act } from 'react-dom/test-utils';

export const wait = (timeout = 0) =>
  new Promise((resolve) => {
    setTimeout(resolve, timeout);
  });

export const actWait = async (timeout = 0) => {
  await act(async () => {
    await wait(timeout);
  });
};
