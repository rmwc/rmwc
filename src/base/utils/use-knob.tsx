/* istanbul ignore file */
import { useState } from 'react';
import * as knobTypes from '@storybook/addon-knobs';

// @ts-ignore
import { manager } from '@storybook/addon-knobs/dist/esm/registerKnobs';

export const useKnob = (
  knobType: keyof typeof knobTypes,
  name: string,
  defaultValue: any
) => {
  const knobFunc = knobTypes[knobType] as any;
  const [stateValue, _stateSetter] = useState(defaultValue);
  const knobValue = knobFunc(name, stateValue);

  const stateSetter = (value: any) => {
    _stateSetter(value);
    manager.knob(name, stateValue);
  };

  if (knobValue !== stateValue) {
    stateSetter(knobValue);
  }
  return [stateValue, stateSetter];
};
