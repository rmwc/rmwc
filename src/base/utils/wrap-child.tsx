import React from 'react';
import classNames from 'classnames';

export const wrapChild = (props: { [key: string]: any }) => {
  const child = React.Children.only(props.children);
  return React.cloneElement(child, {
    ...props,
    ...child.props,
    className: classNames(props.className, child.props.className),
    style: { ...child.props.style, ...props.style }
  });
};
