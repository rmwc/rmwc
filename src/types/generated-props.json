{
  "npm": {},
  "typescript": {
    "WithRippleProps": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "interface",
      "name": "WithRippleProps",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L46",
      "methods": [],
      "properties": [
        {
          "documentation": {
            "contents": [
              "<p>Adds a ripple effect to the component</p>\n"
            ],
            "contentsRaw": "Adds a ripple effect to the component",
            "metadata": {}
          },
          "fileName": "src/types/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "ripple",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L48",
          "type": "RipplePropT"
        }
      ]
    },
    "IconOptions": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "interface",
      "name": "IconOptions",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L70",
      "indexSignature": {
        "documentation": {
          "contents": [
            "<p>Additional props</p>\n"
          ],
          "contentsRaw": "Additional props",
          "metadata": {}
        },
        "fileName": "src/types/index.tsx",
        "kind": "signature",
        "name": "__index",
        "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L97",
        "parameters": [
          {
            "flags": {
              "isExported": false,
              "isExternal": false,
              "isOptional": false,
              "isPrivate": false,
              "isProtected": false,
              "isPublic": false,
              "isRest": false,
              "isStatic": false
            },
            "kind": "parameter",
            "name": "key",
            "type": "string"
          }
        ],
        "returnType": "any",
        "type": "{ [key: string]: any }"
      },
      "methods": [],
      "properties": [
        {
          "documentation": {
            "contents": [
              "<p>A base className for the icon namespace, i.e. material-icons.</p>\n"
            ],
            "contentsRaw": "A base className for the icon namespace, i.e. material-icons.",
            "metadata": {}
          },
          "fileName": "src/types/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "basename",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L89",
          "type": "undefined | string"
        },
        {
          "fileName": "src/types/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": false,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "icon",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L71",
          "type": "IconElementT"
        },
        {
          "documentation": {
            "contents": [
              "<p>A className prefix to use when using css font icons that use prefixes,\ni.e. font-awesome-, ion-, glyphicons-.\nThis only applies when using the &#39;className&#39; strategy.</p>\n"
            ],
            "contentsRaw": "A className prefix to use when using css font icons that use prefixes,\ni.e. font-awesome-, ion-, glyphicons-.\nThis only applies when using the 'className' strategy.",
            "metadata": {}
          },
          "fileName": "src/types/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "prefix",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L87",
          "type": "undefined | string"
        },
        {
          "documentation": {
            "contents": [
              "<p>A render function to use when using the &#39;custom&#39; strategy.</p>\n"
            ],
            "contentsRaw": "A render function to use when using the 'custom' strategy.",
            "metadata": {}
          },
          "fileName": "src/types/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "render",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L91",
          "type": "undefined | (props: ) => Exclude<React.ReactNode, React.ReactText>"
        },
        {
          "documentation": {
            "contents": [
              "<p>A size to render the icon</p>\n"
            ],
            "contentsRaw": "A size to render the icon",
            "metadata": {}
          },
          "fileName": "src/types/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "size",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L97",
          "type": "IconSizeT"
        },
        {
          "documentation": {
            "contents": [
              "<p>Handle multiple methods of embedding an icon.\n&#39;ligature&#39; uses ligature style embedding like material-icons,\n&#39;className&#39; adds a class onto the element for libraries like glyphicons and ion icons,\n&#39;url&#39; will load a remote image, and\n&#39;component&#39; will render content as children like SVGs or any other React node.\n&#39;custom&#39; allows you to specify your own render prop.\nIf not set, &#39;auto&#39; will be used or the defaults set inside of RMWCProvider.</p>\n"
            ],
            "contentsRaw": "Handle multiple methods of embedding an icon.\n'ligature' uses ligature style embedding like material-icons,\n'className' adds a class onto the element for libraries like glyphicons and ion icons,\n'url' will load a remote image, and\n'component' will render content as children like SVGs or any other React node.\n'custom' allows you to specify your own render prop.\nIf not set, 'auto' will be used or the defaults set inside of RMWCProvider.",
            "metadata": {}
          },
          "fileName": "src/types/index.tsx",
          "flags": {
            "isExported": true,
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "strategy",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L81",
          "type": "IconStrategyT"
        }
      ]
    },
    "ThemeOptionT": {
      "documentation": {
        "contents": [
          "<p>Theming</p>\n"
        ],
        "contentsRaw": "Theming",
        "metadata": {}
      },
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "ThemeOptionT",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L4",
      "type": "\"primary\" | \"secondary\" | \"background\" | \"surface\" | \"error\" | \"primaryBg\" | \"secondaryBg\" | \"onPrimary\" | \"onSecondary\" | \"onSurface\" | \"onError\" | \"textPrimaryOnBackground\" | \"textSecondaryOnBackground\" | \"textHintOnBackground\" | \"textDisabledOnBackground\" | \"textIconOnBackground\" | \"textPrimaryOnLight\" | \"textSecondaryOnLight\" | \"textHintOnLight\" | \"textDisabledOnLight\" | \"textIconOnLight\" | \"textPrimaryOnDark\" | \"textSecondaryOnDark\" | \"textHintOnDark\" | \"textDisabledOnDark\" | \"textIconOnDark\" | undefined"
    },
    "ThemePropT": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "ThemePropT",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L33",
      "type": "ThemeOptionT | ThemeOptionT[]"
    },
    "RipplePropT": {
      "documentation": {
        "contents": [
          "<p>Ripples</p>\n"
        ],
        "contentsRaw": "Ripples",
        "metadata": {}
      },
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "RipplePropT",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L38",
      "type": "boolean | "
    },
    "TagT": {
      "documentation": {
        "contents": [
          "<p>Components</p>\n"
        ],
        "contentsRaw": "Components",
        "metadata": {}
      },
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "TagT",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L54",
      "type": "string | React.ComponentType<any>"
    },
    "CustomEventT": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "CustomEventT",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L56",
      "type": "CustomEvent<T> & SyntheticEvent<EventTarget>"
    },
    "IconSizeT": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "IconSizeT",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L60",
      "type": "\"xsmall\" | \"small\" | \"medium\" | \"large\" | \"xlarge\""
    },
    "IconStrategyT": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "IconStrategyT",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L62",
      "type": "\"auto\" | \"ligature\" | \"className\" | \"url\" | \"component\" | \"custom\""
    },
    "IconPropT": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "IconPropT",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L102",
      "type": "IconElementT | IconOptions"
    },
    "HTMLProps": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "HTMLProps",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L104",
      "type": "A & ClassAttributes<T> & "
    },
    "ComponentProps": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "ComponentProps",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L111",
      "type": "Props & ElementProps | ComponentPropsWithRef<Tag> & "
    },
    "ComponentType": {
      "fileName": "src/types/index.tsx",
      "flags": {
        "isExported": true,
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "type alias",
      "name": "ComponentType",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/types/index.tsx#L124",
      "type": "(props: ComponentProps<Props, ElementProps, Tag>, ref: any) => Element"
    }
  }
}
