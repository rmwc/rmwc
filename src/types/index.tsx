/**
 * Theming
 */
export type ThemeOptionT =
  | 'primary'
  | 'secondary'
  | 'background'
  | 'surface'
  | 'error'
  | 'primaryBg'
  | 'secondaryBg'
  | 'onPrimary'
  | 'onSecondary'
  | 'onSurface'
  | 'onError'
  | 'textPrimaryOnBackground'
  | 'textSecondaryOnBackground'
  | 'textHintOnBackground'
  | 'textDisabledOnBackground'
  | 'textIconOnBackground'
  | 'textPrimaryOnLight'
  | 'textSecondaryOnLight'
  | 'textHintOnLight'
  | 'textDisabledOnLight'
  | 'textIconOnLight'
  | 'textPrimaryOnDark'
  | 'textSecondaryOnDark'
  | 'textHintOnDark'
  | 'textDisabledOnDark'
  | 'textIconOnDark'
  | undefined;

export type ThemePropT = ThemeOptionT | ThemeOptionT[];

/**
 * Ripples
 */
export type RipplePropT =
  | boolean
  | {
      accent?: boolean;
      surface?: boolean;
      unbounded?: boolean;
    };

export interface WithRippleProps {
  /** Adds a ripple effect to the component */
  ripple?: RipplePropT;
}

/**
 * Components
 */
export type TagT = string | React.ComponentType<any>;

export type CustomEventT<T> = CustomEvent<T> &
  React.SyntheticEvent<EventTarget>;

type IconElementT = React.ReactNode;
export type IconSizeT = 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';

export type IconStrategyT =
  | 'auto'
  | 'ligature'
  | 'className'
  | 'url'
  | 'component'
  | 'custom';

export interface IconOptions {
  icon: IconElementT;
  /**
   * Handle multiple methods of embedding an icon.
   * 'ligature' uses ligature style embedding like material-icons,
   * 'className' adds a class onto the element for libraries like glyphicons and ion icons,
   * 'url' will load a remote image, and
   * 'component' will render content as children like SVGs or any other React node.
   * 'custom' allows you to specify your own render prop.
   * If not set, 'auto' will be used or the defaults set inside of RMWCProvider.
   * */
  strategy?: IconStrategyT;
  /**
   * A className prefix to use when using css font icons that use prefixes,
   * i.e. font-awesome-, ion-, glyphicons-.
   * This only applies when using the 'className' strategy.
   **/
  prefix?: string;
  /** A base className for the icon namespace, i.e. material-icons. */
  basename?: string;
  /** A render function to use when using the 'custom' strategy. */
  render?: (props: {
    content: IconElementT;
    className: string;
    ref?: React.Ref<any>;
  }) => Exclude<React.ReactNode, React.ReactText>; // Don't allow string | number, those would be rendered as {null} anyway
  /** A size to render the icon  */
  size?: IconSizeT;
  /** Additional props */
  [key: string]: any;
}

export type IconPropT = IconElementT | IconOptions;

export type HTMLProps<T = HTMLElement, A = React.AllHTMLAttributes<T>> = A &
  React.ClassAttributes<T> & {
    tag?: TagT;
    theme?: ThemePropT;
    ref?: React.HTMLProps<T>['ref'];
  };

export type ComponentProps<
  Props extends {},
  ElementProps extends {},
  Tag extends React.ElementType
> = Props &
  (
    | ElementProps
    | (React.ComponentPropsWithRef<Tag> & {
        tag?: Tag;
        theme?: ThemePropT;
      })
  );

export type ComponentType<
  Props extends {},
  ElementProps extends {},
  Element extends React.ElementType<any>
> = {
  <Tag extends React.ElementType<any> = Element>(
    props: ComponentProps<Props, ElementProps, Tag>,
    ref: any
  ): JSX.Element;
  displayName?: string;
};
