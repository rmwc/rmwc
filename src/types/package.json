{
  "name": "@rmwc/types",
  "version": "8.0.3",
  "description": "RMWC Shared Typescript Types",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "module": "next/index.js",
  "sideEffects": [
    "./styles.tsx",
    "./styles.js",
    "*.css"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "publish:dist": "npm publish dist --access=public",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/jamesmfriedman/rmwc.git"
  },
  "keywords": [
    "rmwc"
  ],
  "author": "James Friedman",
  "private": false,
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/jamesmfriedman/rmwc/issues"
  },
  "homepage": "https://github.com/jamesmfriedman/rmwc/tree/master/src/types#readme",
  "gitHead": "bf098adc2571ee1641ef20d79f38fb64e9ff186e"
}
