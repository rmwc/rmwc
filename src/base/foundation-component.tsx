// eslint-disable-next-line @typescript-eslint/no-unused-vars
import * as RMWC from '@rmwc/types';
import { SpecificEventListener } from '@material/base/types';
import React, { useEffect, useState, useRef, useMemo } from 'react';
//@ts-ignore
import classNames from 'classnames';
import { eventsMap } from './utils/events-map';
import { toCamel } from './utils/strings';
import { MDCFoundation } from '@material/base';
import { handleRef } from './component';

const reactPropFromEventName = (evtName: string) =>
  (eventsMap as { [key: string]: string })[evtName] || evtName;

export class FoundationElement<Props extends {}, ElementType = HTMLElement> {
  private _classes = new Set<string>();
  private _events: { [key: string]: (evt: Event) => void } = {};
  private _style: { [key: string]: string | number | null } = {};
  private _props: Partial<Props> = {};
  _onChange: (() => void) | null = null;

  constructor(onChange: () => void) {
    this._onChange = onChange;
    this.onChange = this.onChange.bind(this);
    this.addClass = this.addClass.bind(this);
    this.removeClass = this.removeClass.bind(this);
    this.hasClass = this.hasClass.bind(this);
    this.setProp = this.setProp.bind(this);
    this.getProp = this.getProp.bind(this);
    this.removeProp = this.removeProp.bind(this);
    this.setStyle = this.setStyle.bind(this);
    this.addEventListener = this.addEventListener.bind(this);
    this.removeEventListener = this.removeEventListener.bind(this);
  }

  onChange() {
    this._onChange && this._onChange();
  }

  destroy() {
    this._onChange = null;
    this._events = {};
    this._style = {};
    this._props = {};
    this._classes = new Set();
  }

  /**************************************************
   * Classes
   **************************************************/
  addClass(className: string) {
    if (!this._classes.has(className)) {
      this._classes.add(className);
      this.onChange();
    }
  }

  removeClass(className: string) {
    if (this._classes.has(className)) {
      this._classes.delete(className);
      this.onChange();
    }
  }

  hasClass(className: string) {
    return this._classes.has(className);
  }

  /**************************************************
   * Props
   **************************************************/
  setProp(propName: keyof Props, value: any, silent: boolean = false) {
    if (this._props[propName] !== value) {
      this._props[propName] = value;
      !silent && this.onChange();
    }
  }

  getProp(propName: keyof Props) {
    return this._props[propName];
  }

  removeProp(propName: keyof Props) {
    if (this._props[propName] !== undefined) {
      delete this._props[propName];
      this.onChange();
    }
  }

  props(propsToMerge: { [key: string]: any }) {
    const { className = '', style = {} } = propsToMerge;

    // handle merging events
    // the foundation should be able to pass something onClick as well as a user
    // This wraps them in a function that calls both
    const mergedEvents = Object.entries(propsToMerge).reduce(
      (acc: any, [key, possibleCallback]) => {
        const existingCallback = this._events[key];
        if (
          typeof possibleCallback === 'function' &&
          typeof existingCallback === 'function'
        ) {
          const wrappedCallback = (evt: any) => {
            existingCallback(evt);
            return possibleCallback(evt);
          };

          acc[key] = wrappedCallback;
        }
        return acc;
      },
      { ...this._events }
    );

    // handle className
    const mergedClasses = classNames(className, [...this._classes]);

    // handle styles
    const mergedStyles = {
      ...this._style,
      ...style
    };

    return {
      ...propsToMerge,
      ...this._props,
      ...mergedEvents,
      style: mergedStyles,
      className: mergedClasses
    };
  }

  /**************************************************
   * Styles
   **************************************************/
  setStyle(propertyName: string, value: number | string | null) {
    propertyName = propertyName.startsWith('--')
      ? propertyName
      : toCamel(propertyName);

    if (this._style[propertyName] !== value) {
      this._style[propertyName] = value;
      this.onChange();
    }
  }

  /**************************************************
   * Events
   **************************************************/
  addEventListener(evtName: string, callback: SpecificEventListener<any>) {
    const propName = reactPropFromEventName(evtName);
    if (this._events[propName] !== callback) {
      this._events[propName] = callback;
      this.onChange();
    }
  }

  removeEventListener(evtName: string, callback: SpecificEventListener<any>) {
    const propName = reactPropFromEventName(evtName);
    if (this._events[propName]) {
      delete this._events[propName];
      this.onChange();
    }
  }

  /**************************************************
   * Refs
   **************************************************/
  readonly reactRef = React.createRef<HTMLDivElement>();
  get ref(): HTMLDivElement | null {
    return this.reactRef.current;
  }
}

const emitFactory =
  (props: { [key: string]: any }) =>
  (evtType: string, evtData: any, shouldBubble: boolean = false) => {
    let evt;

    evt = new CustomEvent(evtType, {
      detail: evtData,
      bubbles: shouldBubble
    });

    // bugfix for events coming from form elements
    // and also fits with reacts form pattern better...
    // This should always otherwise be null since there is no target
    // for Custom Events
    Object.defineProperty(evt, 'target', {
      value: evtData,
      writable: false
    });

    Object.defineProperty(evt, 'currentTarget', {
      value: evtData,
      writable: false
    });

    // Custom handling for React
    const propName = evtType;

    props[propName] && props[propName](evt);

    return evt;
  };

export const useFoundation = <
  Foundation extends MDCFoundation<any>,
  Elements extends { [key: string]: true },
  Api extends (
    params: {
      [key in keyof Elements]: FoundationElement<Props, HTMLElement>;
    } & { foundation: any }
  ) => any,
  Props extends {
    [key: string]: any;
    foundationRef?: React.Ref<Foundation | null>;
    apiRef?: (ref: ReturnType<Api> | null) => void;
  }
>({
  foundation: foundationFactory,
  props: inputProps,
  elements: elementsInput,
  api
}: {
  foundation: (
    elements: {
      [key in keyof Elements]: FoundationElement<Props, HTMLElement>;
    } & {
      getProps: () => Props;
      emit: (
        evtType: string,
        evtData: any,
        shouldBubble?: boolean
      ) => CustomEvent<any>;
    }
  ) => Foundation;
  props: Props;
  elements: Elements;
  api?: Api;
}) => {
  const [, setIteration] = useState(0);

  const props = useRef(inputProps);
  props.current = inputProps;

  const elements = useMemo(
    () =>
      Object.keys(elementsInput).reduce<
        {
          [key in keyof Elements]: FoundationElement<any, HTMLElement>;
        }
      >((acc, key: keyof Elements) => {
        acc[key] = new FoundationElement<Props, HTMLElement>(() => {
          setIteration((val) => val + 1);
        });
        return acc;
      }, {} as any),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  const foundation = useMemo(() => {
    // init foundation
    const f = foundationFactory({
      ...elements,
      getProps: () => props.current,
      emit: (...args) => emitFactory(props.current)(...args)
    });

    // handle apiRefs
    api && handleRef(props.current.apiRef, api({ foundation: f, ...elements }));

    return f;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    const f = foundation;
    f.init();
    api && handleRef(props.current.apiRef, api({ foundation: f, ...elements }));
    handleRef(props.current.foundationRef, f);

    return () => {
      f.destroy();
      handleRef(props.current.apiRef, null);
      handleRef(props.current.foundationRef, null);
      Object.values(elements).map((element) => element.destroy());
      // @ts-ignore
      props.current = {};
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [foundation, elements]);

  return { foundation: foundation, ...elements };
};
