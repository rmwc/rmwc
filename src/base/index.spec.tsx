import * as RMWC from '@rmwc/types';
import React from 'react';
import { mount } from 'enzyme';
import {
  withTheme,
  randomId,
  wrapChild,
  debounce,
  toCamel,
  toDashCase,
  closest
} from './';
import { FoundationElement } from './foundation-component';
import { wait } from './utils/test-utils';

jest.spyOn(console, 'warn');

describe('RMWC', () => {
  it('works', () => {});
});

describe('FoundationElement', () => {
  it('FoundationElement: handles classNames', () => {
    const inst = {
      root: new FoundationElement(() => {})
    };

    inst.root.addClass('foo');
    inst.root.addClass('foo');
    expect(inst.root.props({}).className.includes('foo')).toBe(true);
    expect(inst.root.hasClass('foo')).toBe(true);

    inst.root.removeClass('foo');
    inst.root.removeClass('foo');
    expect(inst.root.props({}).className.includes('foo')).toBe(false);
    expect(inst.root.hasClass('foo')).toBe(false);
  });

  it('FoundationElement: handles ref', () => {
    const inst = {
      root: new FoundationElement(() => {})
    };

    mount(<div ref={inst.root.setRef} />);

    expect(inst.root.ref instanceof HTMLDivElement).toBe(true);
  });

  it('FoundationElement: handles addEventListener / removeEventListener', () => {
    const inst = {
      root: new FoundationElement(() => {})
    };

    // double up on the adds to check that it doesnt add a second time
    const changeHandler = () => {};
    inst.root.addEventListener('change', changeHandler);
    inst.root.addEventListener('change', changeHandler);
    expect(inst.root.props({}).onChange).toBe(changeHandler);

    // double up on the removes to check that it doesnt remove a second time
    inst.root.removeEventListener('change', changeHandler);
    inst.root.removeEventListener('change', changeHandler);
    expect(inst.root.props({}).onChange).toBe(undefined);

    // Check to make sure we can add a bogus one
    inst.root.addEventListener('onChange', changeHandler);
  });

  it('FoundationElement: handles setStyle', () => {
    const inst = {
      root: new FoundationElement(() => {})
    };

    inst.root.setStyle('color', 'red');
    inst.root.setStyle('color', 'red');
    inst.root.setStyle('--myvar', 'red');
    expect(inst.root.props({}).style).toEqual({
      color: 'red',
      '--myvar': 'red'
    });

    inst.root.setStyle('color', 'blue');

    expect(inst.root.props({}).style).toEqual({
      color: 'blue',
      '--myvar': 'red'
    });
  });

  it('FoundationElement: handles prop setters / getters', () => {
    const inst = {
      root: new FoundationElement<any, any>(() => {})
    };

    inst.root.setProp('title', 'red');
    inst.root.setProp('title', 'red');
    expect(inst.root.props({}).title).toBe('red');
    expect(inst.root.getProp('title')).toBe('red');
    expect(inst.root.removeProp('title'));
    expect(inst.root.removeProp('title'));
    expect(inst.root.getProp('title')).toBe(undefined);
  });

  it('FoundationElement: handles prop merging', async () => {
    let blueChangeCalled = false;
    let redChangeCalled = false;
    const el = mount(
      <div
        className="blue"
        style={{ background: 'blue' }}
        onChange={() => (blueChangeCalled = true)}
      />
    );
    const inst = {
      root: new FoundationElement<any, any>(() => {})
    };

    inst.root.addClass('red');
    inst.root.setStyle('color', 'red');
    inst.root.addEventListener('change', () => (redChangeCalled = true));
    el.update();

    await wait(100);
    const mergedProps = inst.root.props(el.props());
    mergedProps.onChange();
    expect(mergedProps.className).toBe('blue red');
    expect(mergedProps.style).toEqual({ color: 'red', background: 'blue' });
    expect(blueChangeCalled).toBe(true);
    expect(redChangeCalled).toBe(true);
  });
});

describe('Utils', () => {
  it('randomId', () => {
    // @ts-ignore
    process.env.NODE_ENV = 'production';
    randomId();
    // @ts-ignore
    process.env.NODE_ENV = 'test';
  });

  it('debounce', (done) => {
    let val = 0;
    const foo = () => val++;
    const debouncedFoo = debounce(foo, 100);
    debouncedFoo();
    setTimeout(() => {
      expect(val).toBe(1);
      done();
    }, 150);
  });

  it('closest', () => {
    const parent = document.createElement('div');
    const child = document.createElement('div');
    parent.classList.add('foo');

    parent.appendChild(child);

    expect(closest(child, '.foo')).toBe(parent);
    expect(closest(null, '.foo')).toBe(null);
  });

  it('wrapChild', () => {
    const Foo = (props: any) => {
      return wrapChild({ ...props, className: 'foo' });
    };

    const el = mount(
      <Foo>
        <div className="child" />
      </Foo>
    );

    expect(el.html().includes('foo child')).toBe(true);
  });

  it('toCamel', () => {
    expect(toCamel('test-foo')).toBe('testFoo');
  });

  it('toDashCase', () => {
    expect(toDashCase('testFoo')).toBe('test-foo');
  });
});

describe('withTheme', () => {
  it('works with and without classnames', () => {
    const Component = withTheme(({ ...rest }) => <div {...rest} />);
    const el = mount(<Component className="test" theme="primary" />);
    expect(el.html().includes('test'));

    mount(<Component className="test" />);
    expect(el.html().includes('test'));
  });

  it('works with arrays', () => {
    const Component = withTheme(({ ...rest }) => <div {...rest} />);
    const el = mount(<Component theme={['primary']} />);
    expect(el.html().includes('mdc-theme-primary'));
  });

  it('handles deprecations', () => {
    const Component = withTheme(({ ...rest }) => <div {...rest} />);
    // @ts-ignore
    mount(<Component theme="primary foo" />);
    mount(<Component theme="on-primary" />);
  });
});
