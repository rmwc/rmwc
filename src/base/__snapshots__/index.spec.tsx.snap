// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal renders 1`] = `
<DocumentFragment>
  <div
    id="rmwcPortal"
  />
</DocumentFragment>
`;

exports[`withTheme handles deprecations 1`] = `
<div>
  <div
    class="mdc-theme--primary foo"
  />
</div>
`;

exports[`withTheme handles deprecations 2`] = `
<div>
  <div
    class="mdc-theme--on-primary"
  />
</div>
`;
